ext/datadog_cov/datadog_cov.c

```diff
@@ -161,26 +201,140 @@ static void dd_cov_update_coverage(rb_event_flag_t event, VALUE data, VALUE self
   }
 
-  char *filename_ptr = RSTRING_PTR(filename);
-  // if the current filename is not located under the root, we skip it
-  if (strncmp(dd_cov_data->root, filename_ptr, dd_cov_data->root_len) != 0)
+  record_impacted_file(dd_cov_data, filename);
+}
+
+// Get source location for a given class name
+static VALUE get_source_location(VALUE klass_name)
+{
+  return rb_funcall(rb_cObject, rb_intern("const_source_location"), 1, klass_name);
+}
+
+// Get source location for a given class name and swallow any exceptions
+static VALUE safely_get_source_location(VALUE klass_name)
+{
+  return rescue_nil(get_source_location, klass_name);
+}
+
+// Executed on RUBY_INTERNAL_EVENT_NEWOBJ event and captures the source file for the
+// allocated object's class.
+static void process_newobj_event(VALUE tracepoint_data, void *data)
+{
+  VALUE self = (VALUE)data;
+  struct dd_cov_data *dd_cov_data;
+  TypedData_Get_Struct(self, struct dd_cov_data, &dd_cov_data_type, dd_cov_data);
+
+  // careful: this part of tracepoint is running for each allocated object, make it fast
+  rb_trace_arg_t *tracearg = rb_tracearg_from_tracepoint(tracepoint_data);
+  VALUE new_object = rb_tracearg_object(tracearg);
+
+  enum ruby_value_type type = rb_type(new_object);
+  if (type != RUBY_T_OBJECT && type != RUBY_T_STRUCT && type != RUBY_T_DATA)
   {
     return;
   }
 
-  // if ignored_path is provided and the current filename is located under the ignored_path, we skip it too
-  // this is useful for ignoring bundled gems location
-  if (dd_cov_data->ignored_path_len != 0 && strncmp(dd_cov_data->ignored_path, filename_ptr, dd_cov_data->ignored_path_len) == 0)
+  VALUE klass = rb_class_of(new_object);
+  if (klass == Qnil || klass == 0)
   {
     return;
   }
 
-  rb_hash_aset(dd_cov_data->coverage, filename, Qtrue);
+  VALUE klass_name = rb_class_name(klass);
+  if (klass_name == Qnil)
+  {
+    return;
+  }
+
+  VALUE class_covered = rb_hash_aref(dd_cov_data->classes_covered_by_allocation, klass_name);
+  if (class_covered == Qtrue)
+  {
+    return;
+  }
+  rb_hash_aset(dd_cov_data->classes_covered_by_allocation, klass_name, Qtrue);
+
+  // this part of tracepoint is running once per class per test
+
+  // Skip anonymous classes starting with "#<Class".
+  // This small snippet improves performance for worst case by 20% for rubocop test suite: it allows
+  // us to skip the source location lookup that will always fail for anonymous classes.
+  const char *klass_name_ptr = RSTRING_PTR(klass_name);
+  const unsigned long klass_name_len = RSTRING_LEN(klass_name);
+  if (klass_name_len >= 2 && klass_name_ptr[0] == '#' && klass_name_ptr[1] == '<')
+  {
+    return;
+  }
+
+  VALUE source_location = safely_get_source_location(klass_name);
+  if (source_location == Qnil || RARRAY_LEN(source_location) == 0)
+  {
+    return;
+  }
+
+  VALUE filename = RARRAY_AREF(source_location, 0);
+  if (filename == Qnil)
+  {
+    return;
+  }
+
+  record_impacted_file(dd_cov_data, filename);
```

Comment:
As per our discussion of "how much to do inside the tracepoint", have you considered perhaps recording only the observed class (specifically, `klass`), and only separately turning it into a file name? Assuming that folks aren't going to create a gazillion classes per single test case, holding on to all the class objects is probably not a lot of overhead.

Reply:
I think this is a great idea. I would gladly use rb_hash for it: not sure if this part of Ruby API is forbidden to use in trace point too?..

Reply:
I got reports of segfaults because of this approach:

```
/usr/local/ruby/lib/ruby/3.2.0/openssl/ssl.rb:278: [BUG] unexpected situation - recordd:1 current:0
ruby 3.2.3 (2024-01-18 revision 52bb2ac0a6) [x86_64-linux]
```

100% that I need to make the tracepoint code as small as possible, will work on that